tests/pytorch/test_stream.py

```diff
@@ -34,6 +62,64 @@ def test_basics():
     gg = g.to(device=F.ctx())
     OPS.copy_u_sum(gg, xx)
 
+@unittest.skipIf(F._default_context_str == 'cpu', reason="stream only runs on GPU.")
+def test_set_get_stream():
+    current_stream = torch.cuda.current_stream()
+    # test setting another stream
+    s = torch.cuda.Stream(device=F.ctx())
+    torch.cuda.set_stream(s)
+    assert to_dgl_stream_handle(s).value == _dgl_get_stream(to_dgl_context(F.ctx())).value
+    # revert to default stream
+    torch.cuda.set_stream(current_stream)
+
+@unittest.skipIf(F._default_context_str == 'cpu', reason="stream only runs on GPU.")
+def test_record_stream_smoke():
+    # smoke test to ensure it can work
+    stream = torch.cuda.Stream()
+    with torch.cuda.stream(stream):
+        t = nd.array(np.array([1., 2., 3., 4.]), ctx=nd.gpu(0))
+        g = rand_graph(10, 20, device=F.cuda())
+    # NDArray is an internal object, just use DGLStreamHandle
+    t.record_stream(to_dgl_stream_handle(torch.cuda.current_stream()))
+    g.record_stream(torch.cuda.current_stream())
+
+@unittest.skipIf(F._default_context_str == 'cpu', reason="stream only runs on GPU.")
+# borrowed from PyTorch, test/test_cuda.py: test_record_stream()
+def test_record_stream():
+    cycles_per_ms = _get_cycles_per_ms()
+
+    t = nd.array(np.array([1., 2., 3., 4.], dtype=np.float32), ctx=nd.cpu())
+    t.pin_memory_()
+    result = nd.empty([4], ctx=nd.gpu(0))
+    stream = torch.cuda.Stream()
+    ptr = [None]
+
+    # Performs the CPU->GPU copy in a background stream
+    def perform_copy():
+        with torch.cuda.stream(stream):
+            tmp = t.copyto(nd.gpu(0))
+            ptr[0] = F.from_dgl_nd(tmp).data_ptr()
+        torch.cuda.current_stream().wait_stream(stream)
+        tmp.record_stream(
+            to_dgl_stream_handle(torch.cuda.current_stream()))
+        torch.cuda._sleep(int(50 * cycles_per_ms))  # delay the copy
+        result.copyfrom(tmp)
+
+    perform_copy()
+    with torch.cuda.stream(stream):
+        tmp2 = nd.array(np.array([1., 2., 3., 4.], dtype=np.float32), ctx=nd.gpu(0))
+        assert F.from_dgl_nd(tmp2).data_ptr() != ptr[0], 'allocation re-used to soon'
```

Comment:
typo `too soon`

Reply:
I do not understand the logic here. `ptr[0]` holds the location of `tmp`. Shouldn't it be always different than `tmp2`?

Reply:
Imagining we didn't call `.record_stream()` for `tmp`, line 105-106 have launched kernels for sleeping and copying. Then in line 110, from the host side, the lifetime of `tmp` is over, so the caching allocator will reuse the memory block of `tmp` for `tmp2`.

`record_stream` will tell the caching allocator that `torch.cuda.current_stream()` is using `tmp`, please don't free its memory before events in `torch.cuda.current_stream()` have done. So in line 110, the allocator will allocate a new memory block for `tmp2`, and that's what we expect.

Reply:
I see. Thanks.